enhancements/storage/csi-efs-operator.md

```diff
@@ -0,0 +1,335 @@
+   to the community source. The "old" operator is still running.
+2. User edits the `Subscription` and updates the `source:` to the Red Hat certified catalog.
+   (Alternatively, user may un-install the community operator and install the supported one.)
+3. OLM reconciles this change, checks the new catalog and if there is a newer operator there, it will install it as an
+   update. In other words, we need to ship a newer version in the Red Hat catalog than in the community one.
+4. The "new" CSI operator starts. It does not see `ClusterCSIDriver` CR for EFS and only emits Info level alert that
+   there is no CR and does nothing. This means the old CSI driver is still running (installed by the "old" operator)
+   and any apps that use it are still running.
+5. User reads release notes, documentation or alert and creates th CR.
+6. The "new" operator "adopts" the CSI driver objects. Namely, it updates DaemonSet of the CSI driver, which then does
+   rolling update of the driver pods, using supported images.
+
+To adopt the old objects, the new operator must use the same names for all objects created by the old operator.
+
+### [efs-utils](https://github.com/aws/efs-utils)
+
+We need to get efs-utils into the container with the EFS CSI driver to enable data encryption in transport. efs-utils
+are just two Python scripts, which create / monitor stunnel connections between OCP nodes and NFS server somewhere in
+AWS cloud.
+
+We've chosen to use base image approach. There are two reasons: 1) we know the drill, and 2) we don't want
+to support a generic RPM that can be used outside of OCP.
+
+1. Fork `github.com/aws/efs-utils` into `github.com/openshift/aws-efs-utils`.
+2. Create a new base image with the utilities, say `ose-aws-efs-utils-base`:
+   ```Dockerfile
+   FROM: ocp4.8:base
+   RUN yum install stunnel python <and any other deps>
+   COPY <the utilities> /usr/bin
+   ```
+3. The EFS CSI driver then uses it as the base image:
+   ```Dockerfile
+   FROM registry.ci.openshift.org/openshift/release:golang-1.15 AS builder
+   ... build the image ...
+
+   FROM ose-aws-efs-utils-base
+   COPY --from=builder <the driver binary> /usr/bin/
+   ```
+
+It would be nice if we did not ship `ose-aws-efs-utils-base` anywhere, so people can consume the EFS utils only through
+the EFS operator. **We do not want to maintain it for everyone.**
+
+### Code organization
+
+* Reuse existing `github.com/openshift/aws-efs-operator` for the operator.
+  * Move the controller-runtime based operator one to a separate branch (`release-4.7`, even though it contains all
+    previous releases?)
+  * Continue with library-go based one in `master`.
+* Fork the CSI driver into `github.com/openshift/aws-efs-csi-driver`.
+* Fork `github.com/aws/efs-utils` into `github.com/openshift/aws-efs-utils`.
+
+### Open Questions
+
+* How to graduate community operator into a supported one?
```

Comment:
Meaning, importantly, how to prevent someone from installing the community operator into 4.8? Yeah, that's a tough one. Perhaps the supported operator checks for that, refuses to function, and throws an alert?

Reply:
Wouldn't it be enough if we just remove it from OperatorHub?

Reply:
Checking OLM, it supports `minKubeVersion` in CSV files - we may hide the new operator in 4.7 and earlier. Not sure if there is an easy way how to hide the old operator in 4.8 and newer. Added as an extra open item.

Reply:
> Wouldn't it be enough if we just remove it from OperatorHub?

That would prevent pre-4.8 clusters from installing it too, right?